File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/AirProficiencyPresenter.vb
Namespace Controls
    Public Class AirProficiencyPresenter
        Inherits UserControl

        Private ReadOnly _image As New Image
        Public Sub New()
            Content = _image
        End Sub

        Private _proficiency As Integer
        Public Property Proficiency As Integer
            Get
                Return _proficiency
            End Get
            Set(value As Integer)
                If _proficiency <> value Then
                    _proficiency = value
                    If value = 0 Then
                        _image.Source = Nothing
                    ElseIf DesignMode.DesignModeEnabled Then
                        _image.Source = New BitmapImage(New Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/AirProficiency/{value}.png"))
                    Else
                        Dim stringId = "UseItemIcon_Bitmap_" & value
                        Dim source As Object = Nothing
                        If Application.Current.Resources.TryGetValue(stringId, source) Then
                            _image.Source = DirectCast(source, BitmapImage)
                        Else
                            Dim s = New BitmapImage(New Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/AirProficiency/{value}.png"))
                            Application.Current.Resources.Add(stringId, s)
                            _image.Source = s
                        End If
                    End If
                End If
            End Set
        End Property
    End Class
End Namespace

File: src/Core/UWP/Sakuno.ING.Views.UWP.Common/Controls/AirProficiencyPresenter.cs
using System;
using Windows.ApplicationModel;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;
using Windows.UI.Xaml.Media.Imaging;

namespace Sakuno.ING.Views.UWP.Controls
{
    public sealed class AirProficiencyPresenter : UserControl
    {
        private readonly Image _image = new Image();
        public AirProficiencyPresenter() => Content = _image;

        private int _proficiency;
        public int Proficiency
        {
            get => _proficiency;
            set
            {
                if (_proficiency != value)
                {
                    _proficiency = value;
                    if (value == 0)
                        _image.Source = null;
                    else if (DesignMode.DesignModeEnabled)
                        _image.Source = new BitmapImage(new Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/AirProficiency/{value}.png"));
                    else
                    {
                        string stringId = "AirProficiency_Bitmap_" + value;
                        if (Application.Current.Resources.TryGetValue(stringId, out object source))
                            _image.Source = source as BitmapImage;
                        else
                        {
                            var s = new BitmapImage(new Uri($"ms-appx:///Sakuno.ING.Views.UWP.Common/Assets/Images/AirProficiency/{value}.png"));
                            Application.Current.Resources[stringId] = s;
                            _image.Source = s;
                        }
                    }
                }
            }
        }
    }
}
